### PR TITLE
Upgrade CNI plugin to support CNI Spec version 0.3.1

### PR DESCRIPTION
**Pull request by gunjan5**
```bash
vagrant@vagrant:~/go/src/github.com/projectcalico/cni-plugin$ CNI_COMMAND=VERSION ./dist/calico < conf
{"cniVersion":"0.3.0","supportedVersions":["0.1.0","0.2.0","0.3.0"]}
```

TODO left: 
- [ ] add tests for old `0.2.0` and empty `cniVersion`
- [x] update comments
- [ ] update docs manifests to specify `"cniVersion": "0.3.1"` https://github.com/projectcalico/calico/pull/673
- [x] Rev the latest spec version `0.3.1` in if/when this PR gets merged https://github.com/containernetworking/cni/pull/413
- [x] make sure it works without cniVersion specified in the conf

### CODE REVIEW

**Comment by caseydavenport** [comment]
Note that the "documentation" required for this will include updating our manifests / documented CNI configs to specify the version.

File: Makefile [hunk 1]
```diff
@@ -115,15 +115,18 @@ dist/flannel dist/loopback dist/host-local:
 .PHONY: test-containerized
 ## Run the tests in a container (as root)
 test-containerized: run-etcd run-k8s-apiserver build-containerized dist/host-local
+	for cniversion in "0.2.0" "0.3.0" ; do \
```
Comment (tomdee) [inline Makefile]: Did you choose not to do this for the non containerized tests too?

File: glide.yaml [hunk 2]
```diff
@@ -2,7 +2,7 @@ package: github.com/projectcalico/cni-plugin
 import:
 - package: github.com/Sirupsen/logrus
 - package: github.com/containernetworking/cni
-  version: 07a8a28637e97b22eb8dfe710eeae1344f69d16e
+  version: 73c6616946b76b7c467774ae5eb062fafdd477d4
```
Comment (tomdee) [inline glide.yaml]: Why not pin to an actual release?

File: utils/utils.go [hunk 3]
```diff
@@ -100,19 +98,20 @@ func GetIdentifiers(args *skel.CmdArgs) (workloadID string, orchestratorID strin
 	return workloadID, orchestratorID, nil
 }
 
-func PopulateEndpointNets(endpoint *api.WorkloadEndpoint, result *types.Result) error {
-	if result.IP4 == nil && result.IP6 == nil {
+func PopulateEndpointNets(endpoint *api.WorkloadEndpoint, result *current.Result) error {
+
```
Comment (tomdee) [inline utils/utils.go]: no need for a blank line here

File: calico.go [hunk 4]
```diff
@@ -144,12 +149,22 @@ func cmdAdd(args *skel.CmdArgs) error {
...
 			}
 
+			// Convert IPAM result into current Result
+			result, err = current.NewResultFromResult(res)
+			if err != nil {
+				return err
+			}
+
+			if len(result.IPs) == 0 {
+				return goerrors.New("IPAM plugin returned missing IP config")
```
Comment (tomdee) [inline calico.go]: This is a new check right? Why did you add it and why did you use goerrors?

File: k8s/k8s.go [hunk 5]
```diff
@@ -180,11 +182,22 @@ func CmdAddK8s(args *skel.CmdArgs, conf utils.NetConf, nodename string, calicoCl
 		case ipAddrs == "" && ipAddrsNoIpam == "":
 			// Call IPAM plugin if ipAddrsNoIpam or ipAddrs annotation is not present.
 			logger.Debugf("Calling IPAM plugin %s", conf.IPAM.Type)
-			result, err = ipam.ExecAdd(conf.IPAM.Type, args.StdinData)
+			res, err := ipam.ExecAdd(conf.IPAM.Type, args.StdinData)
```
Comment (tomdee) [inline k8s/k8s.go]: Same comment as above about res and result. I think it's confusing having both.

File: k8s/k8s.go [hunk 6]
```diff
@@ -180,11 +182,22 @@ func CmdAddK8s(args *skel.CmdArgs, conf utils.NetConf, nodename string, calicoCl
...
+			logger.Debugf("IPAM plugin returned: %+v", res)
+
+			// Convert IPAM result into current Result
+			result, err = current.NewResultFromResult(res)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(result.IPs) == 0 {
+				return nil, errors.New("IPAM plugin returned missing IP config")
```
Comment (tomdee) [inline k8s/k8s.go]: Different error handling to above.

File: calico.go [hunk 7]
```diff
@@ -144,12 +149,22 @@ func cmdAdd(args *skel.CmdArgs) error {
...
 				"type": conf.IPAM.Type}).Debug("Looking for IPAM plugin in paths")
-			result, err = ipam.ExecAdd(conf.IPAM.Type, args.StdinData)
-			logger.WithField("result", result).Info("Got result from IPAM plugin")
+			res, err := ipam.ExecAdd(conf.IPAM.Type, args.StdinData)
+			logger.WithField("result", res).Info("Got result from IPAM plugin")
 			if err != nil {
 				return err
 			}
 
+			// Convert IPAM result into current Result
```
Comment (gunjan5) [inline calico.go]: update this comment to be more descriptive

File: calico.go (hunk 4 above)
Comment (gunjan5) [inline calico.go]: ill rename goerrors to errors

File: calico_cni_ipam_test.go [hunk 8]
```diff
@@ -25,60 +27,73 @@ var _ = Describe("Calico IPAM Tests", func() {
...
 
-					_, _, exitCode := RunIPAMPlugin(netconf, "DEL", "")
+					_, _, exitCode := RunIPAMPlugin(netconf, "DEL", "", cniVersion)
 					Expect(exitCode).Should(Equal(0))
 				},
 				Entry("IPAM with no configuration", true, false, fmt.Sprintf(`
 			{
-			  "name": "net1",
+			  "cniVersion": "%s",
+				"name": "net1",
```
Comment (tomdee) [inline calico_cni_ipam_test.go]: incorrect indent

File: calico_cni_ipam_test.go [hunk 9]
```diff
@@ -88,17 +103,18 @@ var _ = Describe("Calico IPAM Tests", func() {
 			It("Uses the ipv4 pool", func() {
 				netconf := fmt.Sprintf(`
                 {
-                      "name": "net1",
+                      "cniVersion": "%s",
+											"name": "net1",
```
Comment (tomdee) [inline calico_cni_ipam_test.go]: indent

File: calico_cni_ipam_test.go [hunk 10]
```diff
@@ -161,46 +180,48 @@ var _ = Describe("Calico IPAM Tests", func() {
...
 					    "type": "%s"
 					  }
-					}`, os.Getenv("ETCD_IP"), plugin)
+					}`, cniVersion, os.Getenv("ETCD_IP"), plugin)
 		Context("Pass explicit IP address", func() {
 			It("Return the expected IP", func() {
-				result, _, _ := RunIPAMPlugin(netconf, "ADD", "IP=192.168.123.123")
-				Expect(result.IP4.IP.String()).Should(Equal("192.168.123.123/32"))
+				result, _, _ := RunIPAMPlugin(netconf, "ADD", "IP=192.168.123.123", cniVersion)
+				Expect(result.IPs[0].Address.String()).Should(Equal("192.168.123.123/32"))
```
Comment (tomdee) [inline calico_cni_ipam_test.go]: Should we assert that there is exactly one IP in the result struct?

File: calico_cni_k8s_test.go [hunk 11]
```diff
@@ -351,7 +350,8 @@ var _ = Describe("CalicoCni", func() {
 				It("should successfully assigns the annotated IP address", func() {
 					netconfCalicoIPAM := fmt.Sprintf(`
 				{
-			      "name": "net4",
+			      "cniVersion": "%s",
```
Comment (tomdee) [inline calico_cni_k8s_test.go]: indent

File: Makefile [hunk 12]
```diff
@@ -114,19 +114,20 @@ dist/flannel dist/loopback dist/host-local:
 # .go-pkg-cache can't be used (since tests run as root)
 .PHONY: test-containerized
 ## Run the tests in a container (as root)
-test-containerized: run-etcd run-k8s-apiserver build-containerized dist/host-local
+test-containerized: build-containerized dist/host-local;
```
Comment (tomdee) [inline Makefile]: what is the `;` for?

File: Makefile [hunk 13]
```diff
@@ -114,19 +114,20 @@ dist/flannel dist/loopback dist/host-local:
...
-	-e CNI_SPEC_VERSION=$cniversion \
-	-e LOCAL_USER_ID=0 \
-	-e PLUGIN=calico \
-	-v $(CURDIR):/go/src/github.com/projectcalico/cni-plugin:rw \
-	$(CALICO_BUILD) sh -c '\
-			cd  /go/src/github.com/projectcalico/cni-plugin && \
-			ginkgo'
-	make stop-etcd
-	end
+		make run-etcd; make run-k8s-apiserver; \
```
Comment (tomdee) [inline Makefile]: I don't think this is the right approach. I think creating a new make target called something like test all versions would be better. If you don't agree or understand then lets talk f2f

File: calico_cni_test.go [hunk 14]
```diff
@@ -49,31 +49,34 @@ var _ = Describe("CalicoCni", func() {
...
 
 			It("successfully networks the namespace", func() {
 				containerID, netnspath, session, contVeth, contAddresses, contRoutes, err := CreateContainer(netconf, "", "")
 				Expect(err).ShouldNot(HaveOccurred())
 				Eventually(session).Should(gexec.Exit())
 
-				result := types.Result{}
-				if err := json.Unmarshal(session.Out.Contents(), &result); err != nil {
-					panic(err)
+				result, err := getResultForCurrent(session, cniVersion)
```
Comment (tomdee) [inline calico_cni_test.go]: You should assert the number of IPs you get back

File: calico.go (hunk 4 above)
Comment (gunjan5) [inline calico.go]: actually, nevermind, we have another import from libcalico-go with the same import name `errors` that's why I imported Go's native errors package as `goerrors` (similar to what we do in libcalico-go)

File: k8s/k8s.go (hunk 6 above)
Comment (gunjan5) [inline k8s/k8s.go]: as discussed f2f, it is the same (just different import name because of conflict in calico.go)

File: calico_cni_ipam_test.go (hunk 8 above)
Comment (gunjan5) [inline calico_cni_ipam_test.go]: weird, the indents look fine in my IDE (with same spaces, CRs)

**Comment by gunjan5** [comment]
@tomdee made changes as per your review. PTAL

File: Makefile [hunk 15]
```diff
@@ -119,12 +122,34 @@ test-containerized: run-etcd run-k8s-apiserver build-containerized dist/host-loc
...
+	$(CALICO_BUILD) sh -c '\
+			cd  /go/src/github.com/projectcalico/cni-plugin && \
+			ginkgo'
+	make stop-etcd
+
+## Run the tests in a container (as root) for different CNI spec versions
+## to make sure we don't break backwards compatiblity.
+.PHONY: test-containerized-cni-versions
+test-containerized-cni-versions: build-containerized dist/host-local;
+	for cniversion in "0.2.0" "0.3.1" ; do \
```
Comment (gunjan5) [inline Makefile]: @tomdee I'm running the CI tests with 020 and 031. Should we also run them for 010?

Comment (tomdee) [inline Makefile]: WDYT?

File: calico_cni_ipam_test.go [hunk 16]
```diff
@@ -161,46 +180,53 @@ var _ = Describe("Calico IPAM Tests", func() {
...
+				Expect(len(result.IPs)).Should(Equal(1))
```
Comment (tomdee) [inline calico_cni_ipam_test.go]: The `len` check should be before we try to access the first element...

File: test_utils/utils.go [hunk 17]
```diff
@@ -54,6 +58,35 @@ func WipeEtcd() {
...
+	// Check if the version is older than 0.3.0.
+	// Convert it to 0.3.1 standard if that is the case.
+	if version.Compare(cniVersion, "0.3.0", "<") {
+		r020 := types020.Result{}
+
+		if err := json.Unmarshal(session.Out.Contents(), &r020); err != nil {
+			log.Fatalf("Error unmarshaling session output to Result: %v\n", err)
+		}
+
+		r030, err := current.NewResultFromResult(&r020)
```
Comment (tomdee) [inline test_utils/utils.go]: Why is the variable name r030?

**Comment by tomdee** [comment]
A few more questions and niggles

File: test_utils/utils.go (hunk 17 above)
Comment (gunjan5) [inline test_utils/utils.go]: to represent the result is upconverted to 030 result type (filled with "zero value" for the fields not present in older spec version results)

File: Makefile (hunk 15 above)
Comment (gunjan5) [inline Makefile]: I personally think this is ok since 0.2.0 would be the default if it's not specified, and we're changing all our manifests to use 0.3.1

**Comment by gunjan5** [comment]
updated supported version string:
```bash
CNI_COMMAND=VERSION ./dist/calico < conf
{"cniVersion":"0.3.1","supportedVersions":["0.1.0","0.2.0","0.3.0","0.3.1"]}
```
